streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java

```diff
@@ -149,6 +151,10 @@ public void process(final K key, final V value) {
                     key, context().topic(), context().partition(), context().offset(), context().timestamp(), mergedWindow.start(), mergedWindow.end(), closeTime
                 );
                 lateRecordDropSensor.record();
+                // note: since we add one to the lateness for this reason:
+                // if the window end is equal to the close time, it is considered late.
+                // specifically, it is 1ms late. This equates to the expression we use: (close - end + 1)
```

Comment:
I'm not 100% sure I agree with window end equaling close time makes a record late, but I can see the point, so I'll go with what others think.

Reply:
In this case, it's purely a mathematical statement.

I set the "close time" equal to the first ms in which a record is late, so by definition if window_end==close_time, it is late. Note how the comparison above uses `>`.

I could move the "close time" forward 1ms (adding 1 to the calculation of it above), used `>=` in the comparison, and then what you say would be true, and we'd have no need to add 1 here.

I think either way, we're adding a 1 to some calculation in this method, and we'd need a similar explanatory comment to clarify why.

I'm open to swapping the semantics if you think it's more intuitive the other way.

Reply:
I was thinking of it like, "the hardware store closes at 7pm, so if I show up at 7pm, I can't go in".

Reply:
My thinking was if I walk into a meeting scheduled to start at `1:00:00` precisely at `1:00:00` I'm not late, but again I don't have a firm opinion here, and I can see your point as well.

Reply:
But in this case, we're not talking about the start of the window's lifespan, we're talking about its end.

So the analogy would be more like saying "the meeting starts at `1:00:00` and ends `1hr` later." If you arrive at `1:59:59`, you'll find the meeting wrapping up, but if you arrive at `2:00:00`, you are late, and the meeting is over.

Reply:
Yep, I knew that, but I keep mixing up my examples; maybe I have window dyslexia.

Reply:
Note that for `TimeWindow` as we stated in javadoc ` its end timestamp as exclusive boundary` the current logic that `mergedWindow.end() <= closeTime` is considered late looks good to me.

Regarding whether we should use `closeTime - mergedWindow.end() + 1` or just `closeTime - mergedWindow.end()`, I admit the 1ms seems a bit unnatural but from monitoring purposes I think it may be better aligned with `lateRecordDropSensor`: think, if a users saw `lateRecordDropSensor` is non-zero but `recordLatenessSensor` is 0.